go/client/logclient.go

```diff
@@ -179,20 +180,34 @@ func (c *LogClient) postAndParse(uri string, req interface{}, res interface{}) (
 }
 
 // Attempts to add |chain| to the log, using the api end-point specified by
-// |path|.
-func (c *LogClient) addChainWithRetry(path string, chain []ct.ASN1Cert) (*ct.SignedCertificateTimestamp, error) {
+// |path|. If provided context is passed before submission is complete an
+// error will be returned.
+func (c *LogClient) addChainWithRetry(ctx context.Context, path string, chain []ct.ASN1Cert) (*ct.SignedCertificateTimestamp, error) {
 	var resp addChainResponse
 	var req addChainRequest
 	for _, link := range chain {
 		req.Chain = append(req.Chain, base64.StdEncoding.EncodeToString(link))
 	}
-	done := false
 	httpStatus := "Unknown"
+	backoffSeconds := 0
+	done := false
 	for !done {
-		backoffSeconds := 0
+		backoffTimer := time.NewTimer(time.Duration(backoffSeconds) * time.Second)
```

Comment:
How about moving all of this back off logic into its own function `backoffForRetry(d time.Duration) err` or similar?

In the `ctx->Done()` case it could return `ctx.Err()` so that the caller knew to abort rather than try again.